### PR TITLE
fix(image-optimizer): modern img format not work in old browsers

### LINKED ISSUE #20794: Webp images using next/image are not converted to another format on incompatible browsers
**What version of Next.js are you using?**

10.0.4

**What version of Node.js are you using?**

12.16.0

**What browser are you using?**

Safari <= 13

**What operating system are you using?**

MacOS, iOS

**How are you deploying your application?**

next start, externally hosted assets

**Describe the Bug**

If you pass an image src which is natively a `webp` to the `next/image` component, it is always served as a webp. Safari versions <= 13 do not include webp support so the image fails to load.

The Accept header sent in the request to the `_next/image` endpoint from Safari looks like this:
```
https://github.com/vercel/next.js/blob/356bcdec0302c13af312ddcc5c39e78ef8e40c9d/packages/next/next-server/server/image-optimizer.ts#L234-L240
```

I believe the following code is at fault:
https://github.com/vercel/next.js/blob/356bcdec0302c13af312ddcc5c39e78ef8e40c9d/packages/next/next-server/server/image-optimizer.ts#L234-L240

**Expected Behavior**

On browsers that do not support webp, files should be converted into `jpeg` or `png` (most likely jpeg?). 

**To Reproduce**

- Place a `.webp` in your `public` folder or any other configured public host.
- Use the `next/image` component and set the `src` prop to point to the `.webp` image.
- Observe in Safari or IE or another non-webp supporting browser as the image is served in webp format and fails to render.

Perhaps it would also make sense to add an optional `format(s)` query param which the server accepts, passed through as a prop?

Let me know if you'd like anything else from me, and I'll be happy to lend a hand :)

### PR DESCRIPTION
**Pull request by xyy94813**
Auto convert to best fit media type,
when origin img format is modern media type.
If best fit media type is not support by server use PNG format.

Resolve #20794

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Server Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| buildDuration | 10.3s | 10.3s | ⚠️ +12ms |
| nodeModulesSize | 42.5 MB | 42.5 MB | ⚠️ +937 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.056 | 2.014 | -0.04 |
| / avg req/sec | 1215.74 | 1241.44 | +25.7 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.336 | 1.428 | ⚠️ +0.09 |
| /error-in-render avg req/sec | 1871.66 | 1750.96 | ⚠️ -120.7 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| 677f882d2ed8..854c.js gzip | 13.2 kB | 13.2 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-HASH.js gzip | 6.65 kB | 6.65 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.6 kB | 59.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.3 kB | 31.3 kB | ✓ |
| Overall change | 31.3 kB | 31.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| _app-2a09aa2..4a98.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-8b758..aef6.js gzip | 3.46 kB | 3.46 kB | ✓ |
| amp-e3e9bc99..932c.js gzip | 536 B | 536 B | ✓ |
| hooks-5023ef..3670.js gzip | 888 B | 888 B | ✓ |
| index-3a2b03..c677.js gzip | 227 B | 227 B | ✓ |
| link-177412b..043e.js gzip | 1.67 kB | 1.67 kB | ✓ |
| routerDirect..1a66.js gzip | 303 B | 303 B | ✓ |
| withRouter-9..1b7d.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.66 kB | 8.66 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| _buildManifest.js gzip | 347 B | 347 B | ✓ |
| Overall change | 347 B | 347 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| index.html gzip | 615 B | 615 B | ✓ |
| link.html gzip | 621 B | 621 B | ✓ |
| withRouter.html gzip | 608 B | 608 B | ✓ |
| Overall change | 1.84 kB | 1.84 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Serverless Mode</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| buildDuration | 12s | 12.1s | ⚠️ +112ms |
| nodeModulesSize | 42.5 MB | 42.5 MB | ⚠️ +937 B |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| 677f882d2ed8..854c.js gzip | 13.2 kB | 13.2 kB | ✓ |
| framework.HASH.js gzip | 39 kB | 39 kB | ✓ |
| main-HASH.js gzip | 6.65 kB | 6.65 kB | ✓ |
| webpack-HASH.js gzip | 751 B | 751 B | ✓ |
| Overall change | 59.6 kB | 59.6 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.3 kB | 31.3 kB | ✓ |
| Overall change | 31.3 kB | 31.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| _app-2a09aa2..4a98.js gzip | 1.28 kB | 1.28 kB | ✓ |
| _error-8b758..aef6.js gzip | 3.46 kB | 3.46 kB | ✓ |
| amp-e3e9bc99..932c.js gzip | 536 B | 536 B | ✓ |
| hooks-5023ef..3670.js gzip | 888 B | 888 B | ✓ |
| index-3a2b03..c677.js gzip | 227 B | 227 B | ✓ |
| link-177412b..043e.js gzip | 1.67 kB | 1.67 kB | ✓ |
| routerDirect..1a66.js gzip | 303 B | 303 B | ✓ |
| withRouter-9..1b7d.js gzip | 302 B | 302 B | ✓ |
| Overall change | 8.66 kB | 8.66 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| _buildManifest.js gzip | 347 B | 347 B | ✓ |
| Overall change | 347 B | 347 B | ✓ |

</details>

<details>
<summary><strong>Serverless bundles</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| _error.js | 1.01 MB | 1.01 MB | ✓ |
| 404.html | 2.67 kB | 2.67 kB | ✓ |
| amp.amp.html | 10.4 kB | 10.4 kB | ✓ |
| amp.html | 1.86 kB | 1.86 kB | ✓ |
| hooks.html | 1.92 kB | 1.92 kB | ✓ |
| index.js | 1.01 MB | 1.01 MB | ✓ |
| link.js | 1.06 MB | 1.06 MB | ✓ |
| routerDirect.js | 1.06 MB | 1.06 MB | ✓ |
| withRouter.js | 1.06 MB | 1.06 MB | ✓ |
| Overall change | 5.2 MB | 5.2 MB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 5 Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| buildDuration | 12.3s | 12.1s | -259ms |
| nodeModulesSize | 42.5 MB | 42.5 MB | ⚠️ +937 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.055 | 2.135 | ⚠️ +0.08 |
| / avg req/sec | 1216.8 | 1170.74 | ⚠️ -46.06 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.394 | 1.367 | -0.03 |
| /error-in-render avg req/sec | 1794.04 | 1829.24 | +35.2 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| 597-dc8125e2..ef24.js gzip | 13.1 kB | 13.1 kB | ✓ |
| framework.HASH.js gzip | 39.3 kB | 39.3 kB | ✓ |
| main-HASH.js gzip | 6.6 kB | 6.6 kB | ✓ |
| webpack-HASH.js gzip | 954 B | 954 B | ✓ |
| Overall change | 60 kB | 60 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.1 kB | 31.1 kB | ✓ |
| Overall change | 31.1 kB | 31.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| _app-6e0a903..b885.js gzip | 1.26 kB | 1.26 kB | ✓ |
| _error-9e190..f9d6.js gzip | 3.38 kB | 3.38 kB | ✓ |
| amp-7fd79cc4..99ee.js gzip | 538 B | 538 B | ✓ |
| hooks-f7e199..11b9.js gzip | 904 B | 904 B | ✓ |
| index-3dc22c..ffbb.js gzip | 232 B | 232 B | ✓ |
| link-7cc9e62..ef27.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..323a.js gzip | 308 B | 308 B | ✓ |
| withRouter-a..2ef8.js gzip | 304 B | 304 B | ✓ |
| Overall change | 8.59 kB | 8.59 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| _buildManifest.js gzip | 322 B | 322 B | ✓ |
| Overall change | 322 B | 322 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | xyy94813/next.js converted-modern-format-to-best-fit-format | Change |
| - | - | - | - |
| index.html gzip | 586 B | 586 B | ✓ |
| link.html gzip | 593 B | 593 B | ✓ |
| withRouter.html gzip | 580 B | 580 B | ✓ |
| Overall change | 1.76 kB | 1.76 kB | ✓ |

</details>

#### Diffs
<details>
<summary>Diff for <strong>index.html</strong></summary>

```diff
@@ -43,7 +43,7 @@
         "props": { "pageProps": {} },
         "page": "/",
         "query": {},
-        "buildId": "2AIOV04OFxUrjflx_a06k",
+        "buildId": "0gxKvM8ld3_D4k5smFCEX",
         "isFallback": false,
         "gip": true
       }
@@ -77,11 +77,11 @@
       async=""
     ></script>
     <script
-      src="/_next/static/2AIOV04OFxUrjflx_a06k/_buildManifest.js"
+      src="/_next/static/0gxKvM8ld3_D4k5smFCEX/_buildManifest.js"
       async=""
     ></script>
     <script
-      src="/_next/static/2AIOV04OFxUrjflx_a06k/_ssgManifest.js"
+      src="/_next/static/0gxKvM8ld3_D4k5smFCEX/_ssgManifest.js"
       async=""
     ></script>
   </body>
```
</details>
<details>
<summary>Diff for <strong>link.html</strong></summary>

```diff
@@ -48,7 +48,7 @@
         "props": { "pageProps": {} },
         "page": "/link",
         "query": {},
-        "buildId": "2AIOV04OFxUrjflx_a06k",
+        "buildId": "0gxKvM8ld3_D4k5smFCEX",
         "isFallback": false,
         "gip": true
       }
@@ -82,11 +82,11 @@
       async=""
     ></script>
     <script
-      src="/_next/static/2AIOV04OFxUrjflx_a06k/_buildManifest.js"
+      src="/_next/static/0gxKvM8ld3_D4k5smFCEX/_buildManifest.js"
       async=""
     ></script>
     <script
-      src="/_next/static/2AIOV04OFxUrjflx_a06k/_ssgManifest.js"
+      src="/_next/static/0gxKvM8ld3_D4k5smFCEX/_ssgManifest.js"
       async=""
     ></script>
   </body>
```
</details>
<details>
<summary>Diff for <strong>withRouter.html</strong></summary>

```diff
@@ -43,7 +43,7 @@
         "props": { "pageProps": {} },
         "page": "/withRouter",
         "query": {},
-        "buildId": "2AIOV04OFxUrjflx_a06k",
+        "buildId": "0gxKvM8ld3_D4k5smFCEX",
         "isFallback": false,
         "gip": true
       }
@@ -77,11 +77,11 @@
       async=""
     ></script>
     <script
-      src="/_next/static/2AIOV04OFxUrjflx_a06k/_buildManifest.js"
+      src="/_next/static/0gxKvM8ld3_D4k5smFCEX/_buildManifest.js"
       async=""
     ></script>
     <script
-      src="/_next/static/2AIOV04OFxUrjflx_a06k/_ssgManifest.js"
+      src="/_next/static/0gxKvM8ld3_D4k5smFCEX/_ssgManifest.js"
       async=""
     ></script>
   </body>
```
</details>
</details>
Commit: 65c3411c41a1be27c5ff6ad17b4db87a0dc3b2d5

**Comment by xyy94813** [comment]
Should i rebase this PR?
I need any feedback about this solution.

File: packages/next/next-server/server/image-optimizer.ts [hunk 1]
```diff
@@ -245,7 +245,16 @@ export async function imageOptimizer(
...
     contentType = mimeType
   } else if (upstreamType?.startsWith('image/') && getExtension(upstreamType)) {
-    contentType = upstreamType
+    // use best fit media type if original image format is modern type
+    if (MODERN_TYPES.includes(upstreamType)) {
+      const bestFitMediaType = mediaType(headers.accept)
+      // use png if best fit media type is not supported by server
+      contentType = [PNG, JPEG].includes(bestFitMediaType)
+        ? PNG
+        : bestFitMediaType
```
Comment (styfle) [inline packages/next/next-server/server/image-optimizer.ts:254]: I think this is backwards. If bestFitMediaType matches it should return that

File: test/integration/image-optimizer/test/index.test.js [hunk 2]
```diff
@@ -150,6 +150,20 @@ function runTests({ w, isDev, domains }) {
...
+    const query = { w, q: 75, url: '/test.webp' }
+    const opts = { headers: { accept } }
+    const res = await fetchViaHTTP(appPort, '/_next/image', query, opts)
+    expect(res.status).toBe(200)
+    expect(res.headers.get('Content-Type')).toContain('image/png')
+    expect(res.headers.get('cache-control')).toBe(
+      'public, max-age=0, must-revalidate'
+    )
+    expect(res.headers.get('etag')).toBeTruthy()
+  })
```
Comment (styfle) [inline test/integration/image-optimizer/test/index.test.js:165]: This needs a few more tests to handle the other conditions:
- webp image with no matching accept header
- webp image with jpg matching accept header

**Comment by timneutkens** [comment]
I'm going through stale open PRs currently, given that https://github.com/vercel/next.js/pull/22326#discussion_r741505238 is not addressed yet I'm going to close this PR, but I'd be happy to review a new PR 🙏

**Comment by xyy94813** [comment]
I just checked the latest code. 
I found that I can solve the compatibility issue by configuring `ImageConfig`.

```js
nextConfig = {
  images: {
     // ... other conf
     format: ['image/webp', 'image/png', 'image/jpeg'],
  }
}
```

But, why only support webP in default configure?

https://github.com/vercel/next.js/blob/7e055dddbb5c686dcb8f07828a6bc95aa079d512/packages/next/server/image-optimizer.ts#L58-L66

https://github.com/vercel/next.js/blob/7e055dddbb5c686dcb8f07828a6bc95aa079d512/packages/next/server/image-config.ts#L26-L35

**Pull request closed without merging**